pubpub/platform
# Restrict stage visibility

## Issue(s) Resolved
Closes https://github.com/pubpub/platform/issues/762

## High-level Explanation of PR
Adds a subquery for filtering stage queries to just the ones the user has permissions for. This means the stage queries now require a `userId`, so this also makes sure that we pass `userId`s throughout where we are querying for stages.

## Test Plan

(some comments surrounded by ❓ I am looking for feedback on!)

* Add a contributor member
* Add an entry in the StageMemberships table to give this member admin privileges to a stage
* Sign in as this new contributor
* Comment out these lines so that the contributor member can see the stage page (these lines will eventually be removed when everything is properly restricted). ❓ should this be removed now, or are there other things we are waiting for ❓ 
https://github.com/pubpub/platform/blob/5b8586d5a2a377593cbbb27963399f29aea7cb30/core/app/c/%5BcommunitySlug%5D/layout.tsx#L38-L42
* Using the stageId the contributor got assigned to , visit `http://localhost:3000/c/{community}/stages/{stageId}`. You should see the page and any pubs in that stage. Other stageIds should still 404

Note: the stage won't show up in `http://localhost:3000/c/{community}/stages` because of
https://github.com/pubpub/platform/blob/5b8586d5a2a377593cbbb27963399f29aea7cb30/core/lib/stages.ts#L54-L63

This returns an empty array since the stage still has a `moveConstraintSource`. ❓ Should I rework this logic to still show the stage even though its source doesn't "exist"? Or should the `moveConstraintSource` not be returned since the user doesn't have access to that stage? ❓ 

## Screenshots (if applicable)

## Notes
